cmds/ash/main.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2012-2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"log"
+	"os"
+	"os/exec"
+
+	"github.com/u-root/u-root/pkg/complete"
```

Comment:
One problem I see is it would be hard to extend the completion rules without modifying the ash binary (or whichever binary the completion package is built into).

For example, consider a go package external to u-root called "git2" is included in the rootfs. The developer of "git2" wants to include completion rules for branch names. So typing "git2 checkout maste<tab>" would complete to "git2 checkout master". The only way to do this is modifying the ash binary which means including rules specific to git2 in the u-root project even though git2 itself is not in the u-root project -- this is unwanted.

Reply:
It might be better for there to be a distinct exectuable called "completion" which takes an uncompleted line in stdin and writes the best completion to stdout. Ash would pass the current line to "completion" upon receiving a TAB and print out the new suffix. If the completion executable cannot find a suitable match, it propagates to completion-argv0, so "completion-git2" in the above example.

This also gives you the option to save space while building an initramfs. If the completion binary is not included, ash will not be able to find it and no completions will be available.

Reply:
It entirely depends on how flexible you want it to be for extending the completion rules. Making completion external to the shell could make it more flexible.

Reply:
I would suggest a new completer, which, given a name, runs it in the way that it returns the set of choices. We could call it HelperCompleter or some such. But it's out of scope at present.